### PR TITLE
p2p, network, bzzeth: p2p protocol handlers to async

### PR DESCRIPTION
**Pull request by pradovic**
This is an idea how to run handlers async in `p2p.Peer` run event loop. This is still in progress so I still did not comment all the code, or refactored it to look better. Main idea was to try to:
1. Make all handlers execution async
2. Exit run loop with appropriate error.
3. Instead of relying on the side `peer.Drop`. If I am right, this disconnects the peer with the actual produced error instead of the subprotocol  with error string from the  handler. I am not sure if this is important or desired. If not, we could simplify the code in the run loop a bit.
4. Separate sync execution needed for handshake and a run loop, as I feel it makes the code more simple. If there is a need for sync run loop, we can easily add this as separate function as well (reusing existing private function in the package)

There is a couple of `todo` comments that I need to investigate more, I would appreciate any input about those :)

todo:
1. Write more isolated test to test and document the behavior of `Run` event loop.
2. Rethink stream graceful shutdown a bit. 
3. This PR does not cover logging errors and making the actual error more nice. If we agree on this proposal, we can continue to making an error more nice and refactor a code to look a bit better (ex. use context cancellation instead of the `running` bool + mutex in protocol.Run, etc ...). Also, if we don't think that there is any benefit from 3., I can reduce the code not to support this.
4. Investigate `handleMsgPauser` in stream, that is used for tests and see how it feels this kind of loop.
5. It appears that `serverCollectBatch` function in stream produces some leaking go routines, should investigate it a bit. It looks like it is happening on a `master` branch as well, but still, might happen more now.

LMKWYT :)

### CODE REVIEW

File: network/retrieval/retrieve_test.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ func TestUnsolicitedChunkDelivery(t *testing.T) {
 		})
 
 	// expect peer disconnection
-	err = tester.TestDisconnected(&p2ptest.Disconnect{Peer: node.ID(), Error: errors.New("subprotocol error")})
+	err = tester.TestDisconnected(&p2ptest.Disconnect{Peer: node.ID(), Error: errors.New("Message handler error: (msg code 0): unsolicited chunk delivery from peer: cannot find ruid")})
```
Comment (pradovic) [inline network/retrieval/retrieve_test.go:175]: This is an example of difference in disconnect errors now that I mentioned in 3. in the description.

**Review by janos — COMMENTED** [review]
I agree on the approach and would like this PR polished.

> 
> 1. Write more isolated test to test and document the behavior of `Run` event loop.

That would be great.

> 2. Rethink stream graceful shutdown a bit.

👍 

> 3. This PR does not cover logging errors and making the actual error more nice. If we agree on this proposal, we can continue to making an error more nice and refactor a code to look a bit better (ex. use context cancellation instead of the `running` bool + mutex in protocol.Run, etc ...). Also, if we don't think that there is any benefit from 3., I can reduce the code not to support this.

It would be great to address errors also, if it could be in a followup PR, to make them smaller, and leave this one with only async protocols. I think that there is a lot of room for error improvements and that it may make changes in this PR harder to reviw.

> 4. Investigate `handleMsgPauser` in stream, that is used for tests and see how it feels this kind of loop.

Pauser would be useful if it would be available on protocols level for testing purposes.

> 5. It appears that `serverCollectBatch` function in stream produces some leaking go routines, should investigate it a bit. It looks like it is happening on a `master` branch as well, but still, might happen more now.

We should investigate this, as serverCollectBatch produces long running goroutines when they are waiting for chunk descriptions from the subscription. They can leak if peer or registry are not closed.

> 
> LMKWYT :)

File: p2p/protocols/protocol.go [hunk 2]
```diff
@@ -222,34 +228,111 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
 	p.Disconnect(p2p.DiscSubprotocolError)
 }
 
+//Shutdown waits for async go routines to finish and returns.
+func (p *Peer) Shutdown() {
+	p.mtx.Lock()
+	p.running = false
+	p.mtx.Unlock()
+
+	p.eg.Wait()
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go]: Yup, nice catch! I left it there just in case if we decide to do shutdown with commented code.  I am not sure if we are going to need that, depends on the graceful shutdown. I will switch to wait group for now then and bring back eg if needed.

File: network/retrieval/retrieve_test.go (hunk 1 above)
Comment (pradovic) [reply network/retrieval/retrieve_test.go:175]: @zelig It appears you were right. The disconnect will eventually happen with `subprotocol error` anyway. But, if we return the specific error from the run loop, instead of doing a `peer.Stop()`, the actual error  will be broadcasted via peer broadcast (which is how it is checked in the `tester.TestDisonnected`). I am still not sure if it is useful to us, but it's still probably more clean.

File: p2p/protocols/protocol.go [hunk 3]
```diff
@@ -288,40 +384,46 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
...
-// handleIncoming(code)
-// is called each cycle of the main forever loop that dispatches incoming messages
-// if this returns an error the loop returns and the peer is disconnected with the error
-// this generic handler
-// * checks message size,
-// * checks for out-of-range message codes,
-// * handles decoding with reflection,
-// * call handlers as callbacks
-func (p *Peer) handleIncoming(handle func(ctx context.Context, msg interface{}) error) error {
+// Receive(code) is a sync call that handles incoming message with provided message handler
```
Comment (zelig) [inline p2p/protocols/protocol.go]: (code) ?

File: p2p/protocols/protocol.go [hunk 4]
```diff
@@ -288,40 +384,46 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
...
+func (p *Peer) Receive(handler func(ctx context.Context, msg interface{}) error) error {
```
Comment (zelig) [inline p2p/protocols/protocol.go]: sure we want this exported?

File: p2p/protocols/protocol.go [hunk 5]
```diff
@@ -288,40 +384,46 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
...
+// * checks for out-of-range message codes,
+// * handles decoding with reflection,
+// * call handlers as callbacks
+func (p *Peer) handleMsg(msg *p2p.Msg, handle func(ctx context.Context, msg interface{}) error) error {
+	// make sure that the payload has been fully consume
 	defer msg.Discard()
 
 	if msg.Size > p.spec.MaxMsgSize {
-		return errorf(ErrMsgTooLong, "%v > %v", msg.Size, p.spec.MaxMsgSize)
+		err := errorf(ErrMsgTooLong, "%v > %v", msg.Size, p.spec.MaxMsgSize)
```
Comment (zelig) [inline p2p/protocols/protocol.go]: why these changes needed?

**Review by zelig — COMMENTED** [review]
Brilliant PR thanks
- many  places logging an error and returning the same error with less context. 
Consider keeping the richer context of the log in the error but only return the error 
- the stream package needs serious cleanup, I put a lot of comments but feel free to keep changes to the minimum and defer sorting it out in  subsequent PRs by whoever takes it up

**Comment by pradovic** [comment]
Thanks @zelig! I agree with both points, I left it for the next PR, as @janos and me wanted to go through errors, especially in stream, after this! I will fix all other suggested changes, and leave error related for the reference for the next PR to avoid to bloat this one too much. I will also add couple of unit tests in this one. Sounds good?

File: bzzeth/bzzeth.go [hunk 6]
```diff
@@ -172,19 +173,24 @@ func (b *BzzEth) handleNewBlockHeaders(ctx context.Context, p *Peer, msg *NewBlo
...
 				p.logger.Debug("bzzeth.handleNewBlockHeaders", "delivered", deliveredCnt)
-				return
+				// todo: introduce better errors
+				return nil
 			}
 			ch := newChunk(hdr)
 			deliveredCnt++
 			p.logger.Trace("bzzeth.handleNewBlockHeaders", "hash", ch.Address().Hex(), "delivered", deliveredCnt)
+
+			req.lock.RLock()
```
[the remark this answers is not in the document]
Comment (pradovic) [reply bzzeth/bzzeth.go:183]: It was async before as well. I am not sure is it really a bug, but it might be potentially dangerous. Basically, if I am right, there is a lock in req that is to be used to update `hashes` map. This lock is used for update but not for read, which might not be fatal, but still a bit not correct :)

File: network/hive.go [hunk 7]
```diff
@@ -390,6 +390,8 @@ func (h *Hive) handlePeersMsg(d *Peer, msg *peersMsg) error {
 func (h *Hive) handleSubPeersMsg(ctx context.Context, d *Peer, msg *subPeersMsg) error {
 	d.setDepth(msg.Depth)
 	// only send peers after the initial subPeersMsg
+	h.lock.Lock()
```
[the remark this answers is not in the document]
Comment (pradovic) [reply network/hive.go:393]: Correct, it was not, it used to be sync. Basically, the check for `sentPeers` boolean value should pass only once, and then all other calls should skip this part. Does it makes sense?

File: network/stream/stream.go [hunk 8]
```diff
@@ -1131,21 +1098,19 @@ func (r *Registry) Stop() error {
...
-	// wait for all handlers to finish
-	done := make(chan struct{})
-	go func() {
-		r.handlersWg.Wait()
-		close(done)
-	}()
-	select {
-	case <-done:
-	case <-time.After(5 * time.Second):
+
```
[the remark this answers is not in the document]
Comment (pradovic) [reply network/stream/stream.go:1063]: This shutdown was/is a bit fuzzy:) If I understood correctly, the old code closed the `quit` channel in registry and was basically looping the messages without executing the  async task. Basically, it is waiting for shutdown to complete until it exits the run loop.  For every execution of the async part, it was increasing the wg handle. Then, on shutdown it was waiting for the wg handle for 5 seconds, to gracefully shutdown, while still relying on some peer.Drop to happen after to cancel to run loop.
In this PR I  have removed the wg handle from the registry, and I moved the shutdown and wait logic into the peer. So, now registry needs to range over all peers in initiate the shutdown and wait for them to wait or timeout, which results in the similar behavior, but a more simple handler, imo. This way, the `quit` channel is not used to actually stop the async part, but to make the handler logic quit appropriately, and the actual Peer.Shutdown is stopping the execution of the async jobs, and also returning `nil` from the run loop which should result in disconnecting the peer, so it shift a bit more control into the protocol.Run. At least this was my rationale. Does it makes sense to you as well? :)

File: p2p/protocols/protocol.go (hunk 5 above)
Comment (pradovic) [reply p2p/protocols/protocol.go]: Oh, sorry, it's probably just my debug leftover 🙈

File: network/retrieval/retrieve.go [hunk 9]
```diff
@@ -340,22 +339,24 @@ func (r *Retrieval) handleRetrieveRequest(ctx context.Context, p *Peer, msg *Ret
 	if err != nil {
 		p.logger.Error("retrieval.handleRetrieveRequest - peer delivery failed", "ref", msg.Addr, "err", err)
 		osp.LogFields(olog.Bool("delivered", false))
-		return
+		// continue in event loop
```
[the remark this answers is not in the document]
Comment (pradovic) [reply network/retrieval/retrieve.go:336]: Here, the error is not returned, to continue the event loop we must return nil, I guess that's why there was a log here before, not to break the loop, just log an error.

File: network/stream/stream.go [hunk 10]
```diff
@@ -690,17 +662,14 @@ func (r *Registry) serverHandleWantedHashes(ctx context.Context, p *Peer, msg *W
 		}
 		// set all chunks as synced
 		if err := provider.Set(ctx, allHashes...); err != nil {
-			p.logger.Error("error setting chunk as synced", "addrs", allHashes, "err", err)
-			p.Drop("error setting chunk as synced")
-			return
+			// todo: handle errors in this function as well
```
[the remark this answers is not in the document]
Comment (pradovic) [reply network/stream/stream.go]: There is a providerSetTimer call below that needs to be done below, if I see it correctly :)

File: p2p/protocols/protocol.go (hunk 3 above)
Comment (pradovic) [reply p2p/protocols/protocol.go]: Some copy-paste mistake 🙈

File: p2p/protocols/protocol.go (hunk 4 above)
Comment (pradovic) [reply p2p/protocols/protocol.go]: Nope, good catch. We can export it if needed, later

File: p2p/protocols/protocol.go [hunk 11]
```diff
@@ -217,39 +225,127 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	p.setRunning(true)
+
+	done := make(chan error, 1)
+	go func() {
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:258]: Nice idea, I will try something like that. We need one channel for shutdown as well. So either use one global channel, or use this local channel + global shutdown channel.

[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:258]: I have reduced the logic to one local channel, and wrapped it into the waits function, which makes the code look similar to this now :) If we need to separate them again when rewriting shutdown logic, we can do it similarly like this as well, with just one  select statement  instead of returning just errc. Thanks for the tip!

File: network/stream/stream.go [hunk 12]
```diff
@@ -612,29 +573,28 @@ func (r *Registry) clientHandleOfferedHashes(ctx context.Context, p *Peer, msg *
...
+		return fmt.Errorf("error sending wanted hashes: %w", err)
 	}
 	if ctr == 0 {
 		// request the next range in case no chunks wanted
-		r.requestSubsequentRange(ctx, p, provider, w, msg.LastIndex)
-		return
+		return r.requestSubsequentRange(ctx, p, provider, w, msg.LastIndex)
+	}
+
+	if errc == nil {
```
Comment (janos) [inline network/stream/stream.go]: This is a very nice catch. But, nil errc should not be an error. In fact, read from nil channel blocks so the select bellow will wait for SyncBatchTimeout, which does not look correct. Should we just return nil in this case?

File: p2p/protocols/protocol.go [hunk 13]
```diff
@@ -225,31 +231,111 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
+					p.mtx.Unlock()
+
+					select {
+					case errc <- err:
+					default:
+					}
+				}
+			}()
+
+			p.mtx.RUnlock()
```
Comment (janos) [inline p2p/protocols/protocol.go]: RUnlock can be called before starting the goroutine as it does not guard it.

File: p2p/protocols/protocol.go [hunk 14]
```diff
@@ -225,31 +231,111 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
-// Drop disconnects a peer.
+func (p *Peer) readMsg() (*p2p.Msg, error) {
+	msg, err := p.rw.ReadMsg()
+	if err != nil {
+		if err != io.EOF {
+			metrics.GetOrRegisterCounter("peer.readMsg.error", nil).Inc(1)
+			log.Error("peer.readMsg", "err", err)
+		}
+	}
+	return &msg, err
```
Comment (janos) [inline p2p/protocols/protocol.go]: p2p.Msg is using value semantics, no need for pointer. This is causing a unneeded dereference in handleMsg.

File: p2p/protocols/protocol_test.go [hunk 15]
```diff
@@ -372,11 +372,163 @@ func TestNoHook(t *testing.T) {
...
+		var wg sync.WaitGroup
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			if err := peer.Run(handler); err != nil {
+				t.Fatal(err)
+			}
+		}()
+
+		time.Sleep(300 * time.Millisecond)
```
Comment (janos) [inline p2p/protocols/protocol_test.go]: Do not use plain time.Sleep for synchronisation. Consider using available concurrency tools or retry functions as the last option. This goes for all new tests in this PR.

File: p2p/protocols/protocol_test.go [hunk 16]
```diff
@@ -372,11 +372,163 @@ func TestNoHook(t *testing.T) {
...
+			return nil
+		}
+
+		var wg sync.WaitGroup
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			if err := peer.Run(handler); err != nil {
+				if err.Error() != "Message handler error: (msg code 0): test error" {
+					t.Fatal(err)
```
Comment (janos) [inline p2p/protocols/protocol_test.go]: Testing Fatal (or Error) function must not be called in a goroutine.

**Review by janos — COMMENTED** [review]
Nice, Peter. This PR needs smoke test validations, as well. We can do that together.

I have a few comments, nothing so significant, except in new tests.

Next step would be to improve errors from protocols, I believe.

File: network/stream/stream.go (hunk 12 above)
Comment (pradovic) [reply network/stream/stream.go]: I agree, I think we should just return nil

File: p2p/protocols/protocol_test.go (hunk 15 above)
Comment (pradovic) [reply p2p/protocols/protocol_test.go]: Sounds good. I added it while debugging, these particular tests can be without it at all.

**Comment by pradovic** [comment]
@janos, @zelig I believe I fixed what was suggested in the comments. Some stuff related to the stream, errors and pause is left for the next error-related PRs, should come very soon. Also, during the error pruning, I will investigate a bit more if need the correct error in the node events, and if we don't need it (no need to propagate errors, other then readMsg errors in the run method), then I will switch back to peer drop. Basically, it will just make our run function less complicated, but we will still log correct errors.

**Comment by pradovic** [comment]
Smoke tests seem to pass. @janos thanks for help!

**Review by janos — COMMENTED** [review]
@pradovic if you could address these two minor comments https://github.com/ethersphere/swarm/pull/2018/files#r357484127 https://github.com/ethersphere/swarm/pull/2018/files#r357485691, otherwise, LGTM.

**Review by janos — APPROVED** [review]

File: network/hive.go [hunk 17]
```diff
@@ -390,24 +390,30 @@ func (h *Hive) handlePeersMsg(d *Peer, msg *peersMsg) error {
...
-				peers = append(peers, p.BzzAddr)
-			}
-			return true
-		})
-		// if useful  peers are found, send them over
-		if len(peers) > 0 {
-			go d.Send(ctx, &peersMsg{Peers: sortPeers(peers)})
-		}
+	h.lock.Lock()
+	if d.sentPeers {
```
Comment (acud) [inline network/hive.go:394]: @pradovic please run all `hive` tests multiple times (with high `count` values). i've made some changes to hive message handlers at the time and get tests to start flaking for some very bizarre reasons (which to date i have not managed to understand why)

File: network/retrieval/retrieve.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (acud) [reply network/retrieval/retrieve.go:336]: @zelig if a chunk is not found in this case an error will be returned, then the peer is dropped as a result (which is incorrect if the chunk cannot be found)

File: network/retrieval/retrieve.go [hunk 18]
```diff
@@ -340,22 +339,24 @@ func (r *Retrieval) handleRetrieveRequest(ctx context.Context, p *Peer, msg *Ret
...
-func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *ChunkDelivery) {
+func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *ChunkDelivery) error {
 	p.logger.Debug("retrieval.handleChunkDelivery", "ref", msg.Addr)
 	err := p.checkRequest(msg.Ruid, msg.Addr)
 	if err != nil {
 		unsolicitedChunkDelivery.Inc(1)
 		p.logger.Error("unsolicited chunk delivery from peer", "ruid", msg.Ruid, "addr", msg.Addr, "err", err)
-		p.Drop("unsolicited chunk delivery")
-		return
+		return fmt.Errorf("unsolicited chunk delivery from peer: %s", err)
```
[the remark this answers is not in the document]
Comment (acud) [reply network/retrieval/retrieve.go:352]: :+1: you can remove `L358`

File: network/retrieval/retrieve.go [hunk 19]
```diff
@@ -386,9 +387,11 @@ func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *Chunk
 	if err != nil {
 		p.logger.Error("netstore error putting chunk to localstore", "err", err)
 		if err == storage.ErrChunkInvalid {
-			p.Drop("invalid chunk in netstore put")
+			return fmt.Errorf("netstore error putting chunk to localstore: %s", err)
```
Comment (acud) [inline network/retrieval/retrieve.go:383]: if the chunk is invalid then we should definitely drop the peer, but if netstore `Put` fails we should definitely consider propagating the error too. @janos? @zelig?

File: p2p/protocols/protocol.go [hunk 20]
```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
+		if err != io.EOF {
+			metrics.GetOrRegisterCounter("peer.readMsg.error", nil).Inc(1)
+			log.Error("peer.readMsg", "err", err)
+		}
+	}
+	return msg, err
+}
+
+// Drop disconnects a peer
 // TODO: may need to implement protocol drop only? don't want to kick off the peer
```
Comment (acud) [inline p2p/protocols/protocol.go]: can also be removed

File: pss/outbox/outbox_test.go [hunk 21]
```diff
@@ -115,14 +114,20 @@ func TestOutboxWorkers(t *testing.T) {
 	var maxParallel int32
 
 	var wg sync.WaitGroup
+	var mtx sync.Mutex
 	mockForwardFunction := func(msg *message.Message) error {
-		atomic.AddInt32(&parallel, 1)
+		mtx.Lock()
```
Comment (acud) [inline pss/outbox/outbox_test.go:119]: why the change? it is more readable with `atomic` IMO

**Review by acud — CHANGES_REQUESTED** [review]

File: network/stream/stream.go [hunk 22]
```diff
@@ -497,16 +480,17 @@ func (r *Registry) serverHandleGetRange(ctx context.Context, p *Peer, msg *GetRa
 		batchSizeGauge.Update(int64(l))
 	}
 	if err := p.Send(ctx, offered); err != nil {
-		p.logger.Error("erroring sending offered hashes", "ruid", msg.Ruid, "err", err)
 		p.mtx.Lock()
 		delete(p.openOffers, msg.Ruid)
 		p.mtx.Unlock()
-		p.Drop("error sending offered hashes")
+		return fmt.Errorf("error sending offered hashes: %w", err)
```
[the remark this answers is not in the document]
Comment (acud) [reply network/stream/stream.go:460]: i chose to delete the entry for the sake of cleanliness, but indeed it is not necessary since the whole peer object should be disposed of when a send error occurs

File: network/stream/stream.go (hunk 12 above)
Comment (acud) [reply network/stream/stream.go]: the only case where `errc` is `nil` is when the `if` statement in line `559` is entered. in that case, the `if` statement in line `576` would be entered too, and the subsequent range would be requested. as a result, it is not possible that this `if` block will ever be triggered so it is safe to remove it IMO

File: network/stream/stream.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (acud) [reply network/stream/stream.go:1063]: makes sense

**Review by acud — CHANGES_REQUESTED** [review]

File: p2p/protocols/protocol.go [hunk 23]
```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
 	}
 }
 
-// Drop disconnects a peer.
+func (p *Peer) readMsg() (p2p.Msg, error) {
+	msg, err := p.rw.ReadMsg()
+	if err != nil {
+		if err != io.EOF {
+			metrics.GetOrRegisterCounter("peer.readMsg.error", nil).Inc(1)
+			log.Error("peer.readMsg", "err", err)
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:313]: I am not sure if we need it, I believe I left it here as it used to be like this before. WDYT?

File: p2p/protocols/protocol.go [hunk 24]
```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
+				return
 			}
-			return err
+
+			p.mtx.RLock()
+			// if loop has been stopped, we don't dispatch any more async routines and discard (consume) the message
+			if !p.running {
+				_ = msg.Discard()
+				p.mtx.RUnlock()
+				continue
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:275]: Again, I did it this way to keep the same logic as before. The loop was still running without executing the active handlers. I guess the reason for this is because the "graceful shutdown" was waiting for async handlers to finish before returning the loop.

File: pss/outbox/outbox_test.go (hunk 21 above)
Comment (pradovic) [reply pss/outbox/outbox_test.go:119]: There was an `atomic.AddInt32(&parallel, 1)` followed by an if check without a lock + both the `atomic.Store()` and regular assignment to assign `maxParallel`. It might not be super important it tests, but I guess that is not thread safe and has a one extra assignment, right?

File: network/hive.go (hunk 17 above)
Comment (pradovic) [reply network/hive.go:394]: Sure, I will, thanks! I have run them multiple times as is, but I will try to increase count values more and see what happens.

Comment (pradovic) [reply network/hive.go:394]: Hey @acud I could not get hive tests to fail. Do you have any particular test and counter  combination in mind? The reason for the lock in this place is that this code is now async.

File: p2p/protocols/protocol.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:313]: Let's tackle this in the error related PR anyway. Is that ok?

File: p2p/protocols/protocol.go [hunk 25]
```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
+			}
+
+			p.mtx.RUnlock()
+			p.wg.Add(1)
+			go func() {
+				defer p.wg.Done()
+				err := p.handleMsg(msg, handler)
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:292]: Yes, I think this is ok. We can expand if we decide to handle multiple runs etc...

File: p2p/protocols/protocol.go [hunk 26]
```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
...
+					case errc <- err:
+					default:
+					}
+				}
+			}()
 		}
+	}()
+
+	return func() error {
+		return <-errc
```
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:304]: Hmmm, the only place where this function is called is in the `run` method, which is waiting for this channel in order to return. How can it leak? I am not sure wdym 🙈 Maybe it could leak if there was multiple callers waiting, but this is not the use-case for now.

File: p2p/protocols/protocol.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (acud) [reply p2p/protocols/protocol.go:313]: cool :+1:

File: p2p/protocols/protocol.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (acud) [reply p2p/protocols/protocol.go:304]: well, if you'll look again you'll see that `errc` is only written to in case of an error, but the channel is never closed otherwise. for cleanliness' sake i'd be happy if there is a `defer close(errc)` in the main goroutine that is on line 249

File: pss/outbox/outbox_test.go (hunk 21 above)
Comment (acud) [reply pss/outbox/outbox_test.go:119]: cool. nice catch

**Review by acud — APPROVED** [review]
LGTM. Please just rebase to resolve conflicts

File: p2p/protocols/protocol.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (pradovic) [reply p2p/protocols/protocol.go:304]: Hmmm, it makes sense in general, but then the "lingering" goroutines can not send to it. It can be solved in there implementation, but since we abandoned this approach in the error related PR I will not update it now, and will keep this in mind if we switch back. Nice catch, thanks!
